File: python/src/memosift/adapters/langchain.py
# LangChain/LangGraph adapter — handles typed message classes and additional_kwargs.
from __future__ import annotations

from typing import TYPE_CHECKING, Any

from memosift.core.pipeline import compress
from memosift.core.types import (
    AnchorLedger,
    MemoSiftMessage,
    ToolCall,
    ToolCallFunction,
)
from memosift.providers.base import LLMResponse, MemoSiftLLMProvider

if TYPE_CHECKING:
    from memosift.config import MemoSiftConfig
    from memosift.core.context_window import ContextWindowState
    from memosift.report import CompressionReport


class LangChainLLMProvider:
    """Wraps any LangChain BaseLanguageModel into MemoSiftLLMProvider."""

    def __init__(self, llm: Any):
        self._llm = llm

    async def generate(
        self,
        prompt: str,
        max_tokens: int = 2048,
        temperature: float = 0.0,
    ) -> LLMResponse:
        """Generate text via LangChain's async invoke.

        Uses ``.bind()`` for model kwargs (not the config dict).
        """
        bound = self._llm.bind(max_tokens=max_tokens, temperature=temperature)
        response = await bound.ainvoke([{"role": "user", "content": prompt}])
        usage = getattr(response, "usage_metadata", None) or {}
        if isinstance(usage, dict):
            in_tokens = usage.get("input_tokens", 0)
            out_tokens = usage.get("output_tokens", 0)
        else:
            in_tokens = getattr(usage, "input_tokens", 0)
            out_tokens = getattr(usage, "output_tokens", 0)
        content = getattr(response, "content", str(response))
        return LLMResponse(text=content, input_tokens=in_tokens, output_tokens=out_tokens)

    async def count_tokens(self, text: str) -> int:
        """Count tokens via LangChain's sync get_num_tokens."""
        try:
            return self._llm.get_num_tokens(text)
        except Exception as e:
            import logging
            import math

            logging.getLogger("memosift.adapters.langchain").debug(
                "LangChain get_num_tokens failed (%s), using heuristic.",
                e,
            )
            return math.ceil(len(text) / 3.5)


def adapt_in(messages: list[Any]) -> list[MemoSiftMessage]:
    """Convert LangChain message objects to MemoSiftMessage list.

    Handles HumanMessage, AIMessage, SystemMessage, ToolMessage and
    their ``additional_kwargs`` / ``response_metadata``.
    """
    result: list[MemoSiftMessage] = []

    for msg in messages:
        # Support both dict-style and LangChain typed messages.
        if isinstance(msg, dict):
            role = msg.get("role", "user")
            content = msg.get("content", "")
            additional_kwargs = msg.get("additional_kwargs", {})
            tool_call_id = msg.get("tool_call_id")
            name = msg.get("name")
        else:
            # LangChain typed message classes.
            msg_type = type(msg).__name__
            role = _langchain_type_to_role(msg_type)
            content = getattr(msg, "content", str(msg))
            additional_kwargs = getattr(msg, "additional_kwargs", {})
            tool_call_id = getattr(msg, "tool_call_id", None)
            name = getattr(msg, "name", None)

        # Extract tool_calls from additional_kwargs (LangChain pattern).
        tool_calls = None
        raw_calls = additional_kwargs.get("tool_calls", [])
        if raw_calls:
            tool_calls = [
                ToolCall(
                    id=tc.get("id", ""),
                    function=ToolCallFunction(
                        name=tc.get("function", {}).get("name", tc.get("name", "")),
                        arguments=tc.get("function", {}).get(
                            "arguments",
                            tc.get("args", "{}"),
                        ),
                    ),
                )
                for tc in raw_calls
            ]

        # Preserve LangChain metadata for round-tripping.
        metadata: dict[str, Any] = {"_langchain_type": type(msg).__name__}
        if additional_kwargs:
            non_tc_kwargs = {k: v for k, v in additional_kwargs.items() if k != "tool_calls"}
            if non_tc_kwargs:
                metadata["_langchain_additional_kwargs"] = non_tc_kwargs
        response_metadata = getattr(msg, "response_metadata", None)
        if response_metadata:
            metadata["_langchain_response_metadata"] = response_metadata

        result.append(
            MemoSiftMessage(
                role=role,
                content=content if isinstance(content, str) else str(content),
                name=name,
                tool_call_id=tool_call_id,
                tool_calls=tool_calls,
                metadata=metadata,
            )
        )

    return result


def adapt_out(messages: list[MemoSiftMessage]) -> list[dict[str, Any]]:
    """Convert MemoSiftMessage list back to LangChain-compatible dicts.

    Returns dicts that can be used with LangChain's message constructors.
    """
    result: list[dict[str, Any]] = []
    for msg in messages:
        d: dict[str, Any] = {
            "role": msg.role,
            "content": msg.content,
        }
        if msg.name is not None:
            d["name"] = msg.name
        if msg.tool_call_id is not None:
            d["tool_call_id"] = msg.tool_call_id

        additional_kwargs: dict[str, Any] = {}
        if msg.tool_calls:
            additional_kwargs["tool_calls"] = [tc.to_dict() for tc in msg.tool_calls]
        # Restore preserved kwargs.
        saved_kwargs = msg.metadata.get("_langchain_additional_kwargs", {})
        additional_kwargs.update(saved_kwargs)
        if additional_kwargs:
            d["additional_kwargs"] = additional_kwargs

        response_metadata = msg.metadata.get("_langchain_response_metadata")
        if response_metadata:
            d["response_metadata"] = response_metadata

        result.append(d)
    return result


async def compress_langchain_messages(
    messages: list[Any],
    *,
    llm: MemoSiftLLMProvider | None = None,
    config: MemoSiftConfig | None = None,
    task: str | None = None,
    ledger: AnchorLedger | None = None,
    context_window: ContextWindowState | None = None,
) -> tuple[list[dict[str, Any]], CompressionReport]:
    """Compress LangChain messages end-to-end.

    Args:
        messages: LangChain message objects or dicts.
        llm: MemoSiftLLMProvider (or use ``LangChainLLMProvider(llm)``).
        config: Pipeline configuration.
        task: Optional task description.
        ledger: Optional AnchorLedger.
        context_window: Context window state for adaptive compression.

    Returns:
        Tuple of (compressed message dicts, compression report).
    """
    memosift_msgs = adapt_in(messages)
    compressed, report = await compress(
        memosift_msgs,
        llm=llm,
        config=config,
        task=task,
        ledger=ledger,
        context_window=context_window,
    )
    return adapt_out(compressed), report


def _langchain_type_to_role(type_name: str) -> str:
    """Map LangChain message class names to roles."""
    mapping = {
        "HumanMessage": "user",
        "AIMessage": "assistant",
        "SystemMessage": "system",
        "ToolMessage": "tool",
        "ChatMessage": "user",
        "FunctionMessage": "tool",
    }
    return mapping.get(type_name, "user")

File: typescript/src/adapters/langchain.ts
// LangChain/LangGraph adapter — handles typed message classes and additional_kwargs.

import type { MemoSiftConfig } from "../config.js";
import { compress } from "../core/pipeline.js";
import type { AnchorLedger, MemoSiftMessage, ToolCall } from "../core/types.js";
import { createMessage } from "../core/types.js";
import type { LLMResponse, MemoSiftLLMProvider } from "../providers/base.js";
import type { CompressionReport } from "../report.js";

// ── LangChain type-name → role mapping ────────────────────────────────────────

const _LANGCHAIN_TYPE_TO_ROLE: Readonly<Record<string, MemoSiftMessage["role"]>> = {
  HumanMessage: "user",
  AIMessage: "assistant",
  SystemMessage: "system",
  ToolMessage: "tool",
  ChatMessage: "user",
  FunctionMessage: "tool",
};

function langchainTypeToRole(typeName: string): MemoSiftMessage["role"] {
  return _LANGCHAIN_TYPE_TO_ROLE[typeName] ?? "user";
}

// ── LLM Provider ──────────────────────────────────────────────────────────────

/**
 * Wraps any LangChain BaseChatModel into a MemoSiftLLMProvider.
 *
 * Uses `.bind()` for model kwargs and `.ainvoke()` (via `invoke` in JS)
 * for generation.
 */
export class LangChainLLMProvider implements MemoSiftLLMProvider {
  private readonly _llm: unknown;

  constructor(llm: unknown) {
    this._llm = llm;
  }

  async generate(
    prompt: string,
    options?: { maxTokens?: number; temperature?: number },
  ): Promise<LLMResponse> {
    const maxTokens = options?.maxTokens ?? 2048;
    const temperature = options?.temperature ?? 0.0;

    // LangChain JS models expose .bind() to set kwargs and .invoke() for async generation.
    const llm = this._llm as Record<string, unknown>;
    const bindFn = llm.bind as (kwargs: Record<string, unknown>) => unknown;
    if (typeof bindFn !== "function") {
      throw new Error("LangChain LLM must expose a .bind() method");
    }
    const bound = bindFn.call(llm, { maxTokens, temperature }) as Record<string, unknown>;

    const invokeFn = bound.invoke as (input: unknown[]) => Promise<unknown>;
    if (typeof invokeFn !== "function") {
      throw new Error("LangChain bound model must expose an .invoke() method");
    }

    const response = (await invokeFn.call(bound, [{ role: "user", content: prompt }])) as Record<
      string,
      unknown
    >;

    // Extract usage metadata.
    const usage = (response.usage_metadata ?? {}) as Record<string, unknown>;
    const inputTokens = typeof usage.input_tokens === "number" ? usage.input_tokens : 0;
    const outputTokens = typeof usage.output_tokens === "number" ? usage.output_tokens : 0;

    const text = typeof response.content === "string" ? response.content : String(response);

    return { text, inputTokens, outputTokens };
  }

  async countTokens(text: string): Promise<number> {
    try {
      const llm = this._llm as Record<string, unknown>;
      const countFn = llm.getNumTokens as (t: string) => number | Promise<number>;
      if (typeof countFn === "function") {
        return await Promise.resolve(countFn.call(llm, text));
      }
    } catch {
      // Fall through to heuristic.
    }
    return Math.ceil(text.length / 3.5);
  }
}

// ── Adapt In ──────────────────────────────────────────────────────────────────

/**
 * Convert LangChain message objects to MemoSiftMessage list.
 *
 * Handles HumanMessage, AIMessage, SystemMessage, ToolMessage and
 * their `additional_kwargs` / `response_metadata`.
 */
export function adaptIn(messages: unknown[]): MemoSiftMessage[] {
  const result: MemoSiftMessage[] = [];

  for (const msg of messages) {
    let role: MemoSiftMessage["role"];
    let content: string;
    let additionalKwargs: Record<string, unknown>;
    let toolCallId: string | null;
    let name: string | null;
    let msgTypeName: string;

    if (msg !== null && typeof msg === "object" && !Array.isArray(msg)) {
      const m = msg as Record<string, unknown>;

      if (typeof m.role === "string" && !m.constructor?.toString().includes("class")) {
        // Dict-style message.
        role = (m.role as MemoSiftMessage["role"]) ?? "user";
        content = typeof m.content === "string" ? m.content : "";
        additionalKwargs = (m.additional_kwargs as Record<string, unknown>) ?? {};
        toolCallId = typeof m.tool_call_id === "string" ? m.tool_call_id : null;
        name = typeof m.name === "string" ? m.name : null;
        msgTypeName = "dict";
      } else {
        // LangChain typed message class.
        const constructor = (msg as { constructor: { name: string } }).constructor;
        msgTypeName = constructor?.name ?? "unknown";
        role = langchainTypeToRole(msgTypeName);
        content = typeof m.content === "string" ? m.content : String(msg);
        additionalKwargs = (m.additional_kwargs as Record<string, unknown>) ?? {};
        toolCallId = typeof m.tool_call_id === "string" ? m.tool_call_id : null;
        name = typeof m.name === "string" ? m.name : null;
      }
    } else {
      // Fallback for unexpected shapes.
      role = "user";
      content = String(msg);
      additionalKwargs = {};
      toolCallId = null;
      name = null;
      msgTypeName = "unknown";
    }

    // Extract tool_calls from additional_kwargs (LangChain pattern).
    let toolCalls: ToolCall[] | null = null;
    const rawCalls = additionalKwargs.tool_calls;
    if (Array.isArray(rawCalls) && rawCalls.length > 0) {
      toolCalls = rawCalls.map((tc: Record<string, unknown>) => {
        const fn = (tc.function ?? {}) as Record<string, unknown>;
        return {
          id: typeof tc.id === "string" ? tc.id : "",
          type: "function",
          function: {
            name:
              typeof fn.name === "string" ? fn.name : typeof tc.name === "string" ? tc.name : "",
            arguments:
              typeof fn.arguments === "string"
                ? fn.arguments
                : typeof tc.args === "string"
                  ? tc.args
                  : "{}",
          },
        };
      });
    }

    // Preserve LangChain metadata for round-tripping.
    const metadata: Record<string, unknown> = {
      _langchain_type: msgTypeName,
    };
    if (Object.keys(additionalKwargs).length > 0) {
      const nonTcKwargs: Record<string, unknown> = {};
      for (const [k, v] of Object.entries(additionalKwargs)) {
        if (k !== "tool_calls") {
          nonTcKwargs[k] = v;
        }
      }
      if (Object.keys(nonTcKwargs).length > 0) {
        metadata._langchain_additional_kwargs = nonTcKwargs;
      }
    }

    const m = msg as Record<string, unknown> | null;
    const responseMetadata =
      m !== null && typeof m === "object"
        ? (m.response_metadata as Record<string, unknown> | undefined)
        : undefined;
    if (responseMetadata && Object.keys(responseMetadata).length > 0) {
      metadata._langchain_response_metadata = responseMetadata;
    }

    result.push(
      createMessage(role, typeof content === "string" ? content : String(content), {
        name,
        toolCallId,
        toolCalls,
        metadata,
      }),
    );
  }

  return result;
}

// ── Adapt Out ─────────────────────────────────────────────────────────────────

/**
 * Convert MemoSiftMessage list back to LangChain-compatible dicts.
 *
 * Returns dicts that can be used with LangChain's message constructors.
 */
export function adaptOut(messages: MemoSiftMessage[]): Record<string, unknown>[] {
  const result: Record<string, unknown>[] = [];

  for (const msg of messages) {
    const d: Record<string, unknown> = {
      role: msg.role,
      content: msg.content,
    };

    if (msg.name != null) {
      d.name = msg.name;
    }
    if (msg.toolCallId != null) {
      d.tool_call_id = msg.toolCallId;
    }

    const additionalKwargs: Record<string, unknown> = {};
    if (msg.toolCalls && msg.toolCalls.length > 0) {
      additionalKwargs.tool_calls = msg.toolCalls.map((tc) => ({
        id: tc.id,
        type: tc.type,
        function: {
          name: tc.function.name,
          arguments: tc.function.arguments,
        },
      }));
    }

    // Restore preserved kwargs.
    const savedKwargs =
      (msg.metadata._langchain_additional_kwargs as Record<string, unknown>) ?? {};
    for (const [k, v] of Object.entries(savedKwargs)) {
      additionalKwargs[k] = v;
    }
    if (Object.keys(additionalKwargs).length > 0) {
      d.additional_kwargs = additionalKwargs;
    }

    const responseMetadata = msg.metadata._langchain_response_metadata as
      | Record<string, unknown>
      | undefined;
    if (responseMetadata) {
      d.response_metadata = responseMetadata;
    }

    result.push(d);
  }

  return result;
}

// ── Convenience function ──────────────────────────────────────────────────────

/**
 * Compress LangChain messages end-to-end.
 *
 * @param messages - LangChain message objects or dicts.
 * @param options - Pipeline options (llm, config, task, ledger).
 * @returns Tuple of compressed message dicts and compression report.
 */
export async function compressLangChainMessages(
  messages: unknown[],
  options?: {
    llm?: MemoSiftLLMProvider | null;
    config?: Partial<MemoSiftConfig> | null;
    task?: string | null;
    ledger?: AnchorLedger | null;
  },
): Promise<[Record<string, unknown>[], CompressionReport]> {
  const memosiftMsgs = adaptIn(messages);
  const { messages: compressed, report } = await compress(memosiftMsgs, {
    llm: options?.llm,
    config: options?.config,
    task: options?.task,
    ledger: options?.ledger,
  });
  return [adaptOut(compressed), report];
}
